STM32F405RGTx/Core/Src/rtos_threads/state_machine.cpp

```diff
@@ -0,0 +1,440 @@
+        return NormalDangerFault;
+    }
+
+    // Check current for phase outputs
+    if (global_mc_data.pIa > MAX_CURRENT_NORMAL || global_mc_data.pIb > MAX_CURRENT_NORMAL || global_mc_data.pIc > MAX_CURRENT_NORMAL) {
+        //TODO: error code
+        return NormalDangerFault;
+    }
+
+    // DC fault checking
+    if (global_mc_data.dc_voltage > MAX_DCVOLTAGE_NORMAL) {
+        // TODO: error code
+        return NormalDangerFault;
+    }
+    else if (global_mc_data.dc_voltage < MIN_DCVOLTAGE_NORMAL) {
+        // TODO: error code
+        return NormalDangerFault;
+    }
+
+    if (global_mc_data.dc_cap_temp > MAX_DCCAP_TEMP_NORMAL) {
+        // TODO: error code
+        return NormalDangerFault;
+    }
+    return NoFault;
+}
+
+State_t StateMachineThread::SevereFaultChecking(void)
+{
+    int overvolt_faults = 0;
+    int undervolt_faults = 0;
+    int temp_faults = 0;
+    // check for temperature faults
+    for (int i = 0; i < NUM_MOSFETS; ++i)
+    {
+        float mosfetTemp = global_mc_data.fet_temps[i];
+        if (mosfetTemp > MAX_MOSFET_TEMP_SEVERE)
+        {
+            ++temp_faults;
+        }
+        if (temp_faults > MIN_TEMP_FAULTS)
+        {
+            // TODO: add error code
+            return SevereDangerFault;
+        }
+    }
+    // check undervolts and overvolts for phase outputs
+    if (global_mc_data.pVa < MIN_VOLTAGE_SEVERE) {
+        undervolt_faults++;
+    }
+    if (global_mc_data.pVb < MIN_VOLTAGE_SEVERE) {
+        undervolt_faults++;
+    }
+    if (global_mc_data.pVc < MIN_VOLTAGE_SEVERE) {
+        undervolt_faults++;
+    }
+
+    if (global_mc_data.pVa > MAX_VOLTAGE_SEVERE) {
+        overvolt_faults++;
+    }
+    if (global_mc_data.pVb > MAX_VOLTAGE_SEVERE) {
+        overvolt_faults++;
+    }
+    if (global_mc_data.pVc > MAX_VOLTAGE_SEVERE) {
+        overvolt_faults++;
+    }
+
+    //  TODO: after checking for faults return state
+    if (overvolt_faults > MIN_OVERVOLT_FAULTS) {
+        return NormalDangerFault;
+    }
+    if (undervolt_faults > MIN_UNDERVOLT_FAULTS) {
+        return NormalDangerFault;
+    }
+
+    // Check current for phase outputs
+    if (global_mc_data.pIa > MAX_CURRENT_SEVERE || global_mc_data.pIb > MAX_CURRENT_SEVERE || global_mc_data.pIc > MAX_CURRENT_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+
+    // DC fault checking
+    if (global_mc_data.dc_voltage > MAX_DCVOLTAGE_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+    else if (global_mc_data.dc_voltage < MIN_DCVOLTAGE_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+    if (global_mc_data.dc_cap_temp > MAX_DCCAP_TEMP_SEVERE) {
+        return SevereDangerFault;
+    }
+    return NoFault;
+}
+
+State_t StateMachineThread::InitializeEvent(void) {
+    return Idle;
+}
+
+State_t StateMachineThread::IdleEvent(void) {
+    SetLedColour(0.0, 50.0, 0.0);
+    State_t normal_fault_check = NormalFaultChecking();
```

Comment:
These fault checking function calls should be moved to the mainloop of this thread to help eliminate some of the repetitiveness (copy and pasting). Again, see the changes that were made to the BMS state machine: https://github.com/waterloop/BMS_SW_G5/pull/15